errors/error.go

```diff
@@ -0,0 +1,373 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors implements the Terramate standard error type.
+// It's heavily influenced by Rob Pike `errors` package in the Upspin project:
+// 	https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
+// See E() for its usage.
+type Error struct {
+	// Kind is the kind of error.
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack StackMeta
+
+	// Err represents the underlying error.
+	Err error
+}
+
+type (
+	// Kind defines the kind of error.
+	Kind string
+	// StackMeta has the metadata of the stack which originated the error.
+	// Same interface as stack.Metadata.
+	StackMeta interface {
+		Name() string
+		Desc() string
+		Path() string
+	}
+)
+
+const separator = ": "
+
+// E builds an error value from its arguments.
+// There must be at least one argument or E panics.
+// The type of each argument determines its meaning. If more than one argument
+// of a given type is presented, only the last one is recorded.
+//
+// The types are:
+// 	errors.Kind
+//		The kind of error (eg.: HCLSyntax, TerramateSchema, etc).
+//	hcl.Range
+//		The file range where the error originated.
+//  errors.Stack
+//		The stack which the error originated.
+//	error
+//		The underlying error that triggered this one.
+//	hcl.Diagnostics
+//		The underlying hcl error that triggered this one.
+//		If this error's FileRange is not set, the diagnostic subject range is
+//		pulled.
+//		If this error's Description is not set, the diagnostic detail field is
+//		pulled.
+//	string
+//		The error description. It supports formatting using the Go's fmt verbs
+//      as long as the arguments are not of the defined types.
+//
+// If the error is printed, only those items that have been
+// set to non-zero values will appear in the result. For the `hcl.Range` type,
+// the `range.Empty()` method is used.
+//
+// Error promotions:
+//
+// If Kind is not specified or Any, we set it to the Kind of
+// the underlying error.
+// If FileRange is not specified, we set it to the FileRange of the underlying
+// error (if set).
+//
+// Minimization:
+//
+// In order to avoid duplicated messages, we erase the fields present in the
+// underlying error if already set with same value in this error.
```

Comment:
```suggestion
// In order to avoid duplicated messages, when wrapping an errors.E inside another errors.E 
// any fields present on the current error overrides the fields defined on the wrapped error.
```

Reply:
overrides is not the correct term because if the underlying error has different field values, we keep them for the error context. We just clear the fields in the underlying error *when* they have the same value in the current error.
I also don't know if saying "wrapping" is a good term because this concept in Go already means wrapping errors with `%w` and this is not supported here.